### PR TITLE
fix(plugins): mixing async and callback style now returns an error

### LINKED ISSUE #5112: throw an error for wrong plugin interfaces
I think once a week I see this error:

```js
app.register(async function plugin (app, opts, next) {
  next()
})
```

I propose to trigger an error when a plugin function has 3 parameters and is an async function.

WDYT? Minor or v5?

### PR DESCRIPTION
**Pull request by giuliowaitforitdavide**
Closes #5112, opened as discussed in #5139

#### Checklist

- [X] run `npm run test` and `npm run benchmark`
- [X] tests and/or benchmarks are included
- [X] documentation is changed or added
- [X] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Review by climba03003 — APPROVED** [review]

File: lib/pluginUtils.js [hunk 1]
```diff
@@ -134,7 +135,14 @@ function registerPluginName (fn) {
   this[kRegisteredPlugins].push(name)
 }
 
+function checkPluginHealthiness (fn) {
```
Comment (metcoder95) [inline lib/pluginUtils.js:138]: Let's just check this gets in sync after #5139 gets merged into `main` 👍

File: lib/errors.js [hunk 2]
```diff
@@ -434,6 +434,12 @@ const codes = {
     'FST_ERR_PLUGIN_NOT_PRESENT_IN_INSTANCE',
     "The decorator '%s'%s is not present in %s"
   ),
+  FST_ERR_PLUGIN_INVALID_ASYNC_HANDLER: createError(
+    'FST_ERR_PLUGIN_INVALID_ASYNC_HANDLER',
+    'Async function has too many arguments. Async plugin should not use the \'done\' argument.',
```
Comment (metcoder95) [inline lib/errors.js:439]:
```suggestion
    'The %s plugin being registered mixes async and callback styles. Async plugin should not mix async and callback style.',
```

File: lib/pluginUtils.js (hunk 1 above)
Comment (giuliowaitforitdavide) [reply lib/pluginUtils.js:138]: Yep, I'm marking the PR as draft

**Review by metcoder95 — APPROVED** [review]
LGTM 👍

**Review by climba03003 — CHANGES_REQUESTED** [review]
Could you rebase your changes based on the latest `next` branch?

**Comment by giuliowaitforitdavide** [comment]
> Could you rebase your changes based on the latest `next` branch?

Done, I also fixed failing tests on the next branch but I'm not sure about the coverage

**Comment by climba03003** [comment]
> Done, I also fixed failing tests on the next branch but I'm not sure about the coverage

Sorry to bother you, I have cleanup the unclean merge from the `next` branch. Could you try again to rebase on top of it?

**Review by climba03003 — APPROVED** [review]
LGTM, I believe the coverage is not related to this PR.
It is related to the `Symbol.asyncDispose`

**Comment by giuliowaitforitdavide** [comment]
> LGTM, I believe the coverage is not related to this PR.
> It is related to the `Symbol.asyncDispose`

Ok, so I won't fix it in this PR. I thought it was my mistake

**Review by Eomm — APPROVED** [review]

**Comment by Eomm** [comment]
Merging - coverage is a `next` issue

**Pull request merged**